src/wasm/wasm-type.cpp

```diff
@@ -950,8 +1030,229 @@ std::ostream& operator<<(std::ostream& os, TypeInfo info) {
     case TypeInfo::RttKind: {
       return os << info.rtt;
     }
+    case TypeInfo::TempRefKind: {
+      os << "(temp ref ";
+      if (info.tempRef.nullable) {
+        os << "null ";
+      }
+      return os << *info.tempRef.heapType << ")";
+    }
+    case TypeInfo::TempRttKind: {
+      return os << "(temp rtt " << info.tempRtt.depth << " "
+                << *info.tempRtt.heapType << ")";
+    }
   }
   WASM_UNREACHABLE("unexpected kind");
 }
 
+struct TypeBuilder::Impl {
+  TypeStore store;
+};
+
+TypeBuilder::TypeBuilder(size_t n) {
+  impl = std::make_unique<TypeBuilder::Impl>();
+  heapTypes.reserve(n);
+  for (; n > 0; --n) {
+    heapTypes.emplace_back(HeapType::InvalidKind);
+  }
+}
+
+TypeBuilder::~TypeBuilder() = default;
+
+Type TypeBuilder::getTempTupleType(const Tuple& tuple) {
+  return impl->store.makeType(tuple);
+}
+
+Type TypeBuilder::getTempRefType(size_t i, bool nullable) {
+  assert(i < heapTypes.size() && "Index out of bounds");
+  return impl->store.makeType(TypeInfo::TempRef{&heapTypes[i], nullable});
+}
+
+Type TypeBuilder::getTempRttType(size_t i, uint32_t depth) {
+  assert(i < heapTypes.size() && "Index out of bounds");
+  return impl->store.makeType(TypeInfo::TempRtt{depth, &heapTypes[i]});
+}
+
+// Implements the algorithm to canonicalize the HeapTypes in a TypeBuilder
+struct Canonicalizer {
```

Comment:
Reading this the first time, I am unclear on what "canonicalization" means. Perhaps a concrete example in a comment could help?

Even reading a second time I am still unsure. I realize that canonicalization is important for recursive types, but as this appears in this PR, it seems it matters in an additional way. My intuition is that without recursion we just allocate N types, if there are N types in the type section, and then set up pointers between them (which is why we need the TypeBuilder). Those pointers could be "final", that is, we allocated N types up front, and those are all of our types. We just need to replace the pointers with copies (but see the comment earlier about pointers), essentially. That seems like it should be possible since each Type provides an indirection (as types are internalized and used by reference). That is, since HeapTypes do not have direct references to each other, the Type indirection gives us all we need. Clearly I'm missing something though :smile: What is it?

Reply:
Maybe the missing piece is that the TypeInfos in the TypeBuilder's TypeStore need to be merged into the globalTypeStore? It can't be as simple as moving the `std::unique_ptr<TypeInfo>`s from one TypeStore to the other because canonical versions of the TypeBuilder's types might already exist in the globalTypeStore. That's why all the temporary types in the type graph need to be replaced with their globally canonical versions.